PokePointer ignore colliders that are not visible

julenka opened the pull request:
## Overview
Following up from #6870 this change additionally allows the **poke pointer** to ignore colliders that are not in the visual FOV.

Since this is a follow up to #6870 which is going into 2.3, I decided to request this go into stabilization. However, this is a risk that this could regress some touch behavior so I would be open to us moving this back to 2.4. Curious to hear other's opinions. I have verified in HandInteractionExample scene and have added tests.

## Changes
- Partial fix #6634. There is one more edge case for line pointers.
- Adds `Camera.IsInFovConeCached` as camera extension to ensure efficient checking of whether colliders are in FOV (as these computations run each frame).
- Adds field to PokePointer specifying whether to ignore colliders that are not in visual FOV. True by default.
- Add tests to verify that poke pointers correctly ignore colliders that aren't in FOV.

## Verification
- Verified that tests pass
- Tested in HandInteractionExample scene


> This optional section is a place where you can detail the specific type of verification 
> you want from reviewers. For example, if you want reviewers to checkout the PR locally
> and validate the functionality of specific scenarios, provide instructions
> on the specific scenarios and what you want verified.
>
> If there are specific areas of concern or question feel free to highlight them here so
> that reviewers can watch out for those issues.
>
> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

julenka commented on Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs:
```diff
@@ -36,10 +37,73 @@ public static bool IsInFOV(this Camera camera, Vector3 position)
         }
 
+
+        // Help to clear caches when new frame runs
```
I'm curious to hear opinions on naming for these static variables, as well as whether people have ideas for how to implement this in a better way.

Troy-Ferrell replied:
I feel like this is complicated enough to be in a utility class by itself?

Other thing that comes up is concurrency. I think....we are fine because Unity restricts all calls to the main thread and this is an extension of Camera. So this should force all calls to IsInFOVConeCached to be serialized and thus not possible for two threads to be accessing the static resources concurrently. Thoughts?

julenka replied:
I tried to think of a good location and name for this, but couldn't really think of anything. I'm thinking again this might be good to refactor once we have other code that does FOV checks where we want caching.

However, to help clarify things I prefixed the variables with inFOVCone so people understand this is related to the fov cone checks.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs:
```diff
@@ -51,6 +52,25 @@ public class PokePointer : BaseControllerPointer, IMixedRealityNearPointer
+        /// a touchable but cannot see it. Visual FOV is defined by cone centered about display center, 
+        /// radius equal to half display height.
+        /// </summary>
+        public bool IgnoreCollidersNotInFOV
```
Maybe? I'm hesitant to modify interfaces since it's hard to undo or change interfaces.

Troy-Ferrell replied:
I agree with Julia, feels more like feature creep for the IMixedRealityPointer interface. 

But it does seem like this would be good in a base class for PokePointer and SpherePointer

julenka replied:
I would prefer not to add more levels of inheritance until there's lots of signal (3 repeated classes) that a field could be re-used. Motivation is I find it really hard to navigate the code in MRTK because many fields are in many levels of base classes. Maybe I just need a better tool for navigating base classes (currently use class diagrams). I am thinking of making a third kind of near pointer, a Cone Pointer, as part of fixing #6634 (which I will probalby break up into multiple bugs). Then it woudl be appropriate to have a base class, I think.

Troy-Ferrell replied:
Very true, base classes impede performance and sometimes code maintainability.

Alternative suggestion...to just consider...because I myself am not sure it's better/good. Why not create a new interface, like IFOVChecker? 

Just a thought

julenka replied:
I'd prefer to reduce the number of new interfaces we add. I think when I build the cone pointer I will probably make a base class or interface, but for now I'll leave it like this...keeps the code simpler

julenka commented:
/azp run

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs:
```diff
@@ -100,52 +100,76 @@ public IEnumerator TestLinePointers()
+                yield return PlayModeTestUtilities.WaitForInputSystemUpdate();
+                Assert.IsTrue(myPointer.IsInteractionEnabled, $"Pointer {myPointer.PointerName} should be enabled.");
+                pokePointer.IgnoreCollidersNotInFOV = true;
+            }
```
This would also be good to simplify with base class shared between the two

julenka replied:
Good point! I still would prefer to wait until we have more need, specifically when creating a ConePointer. I can create a bug for this and mention that following up from #7016 we should make a base class.

Troy-Ferrell replied:
Since you arent' creating an interface or base class, maybe add else statement here that throws error saying test helper isn't valid for this type of pointer? To help notify people who might try to use this in future

Troy-Ferrell commented on Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs:
```diff
@@ -36,10 +37,73 @@ public static bool IsInFOV(this Camera camera, Vector3 position)
+        private static Dictionary<Collider, bool> colliderCache = new Dictionary<Collider, bool>();
+        // List of corners shared across all sphere pointer query instances --
+        // used to store list of corners for a bounds. Shared and static
+        // to avoid allocating memory each frame
```
Remove "Sphere pointer query instances"

// List of corners of bounds box for current collider being processed in IsInFOVConeCached
// Shared and static to avoid allocating additional memory each frame

Troy-Ferrell replied:
Suggestion: I'd also rename to boundsCornerPoints to just be explicit and easier reading

julenka replied:
Renamed!

Troy-Ferrell commented on Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs:
```diff
@@ -36,10 +37,73 @@ public static bool IsInFOV(this Camera camera, Vector3 position)
+                lastCalculatedFrame = Time.frameCount;
+            }
+
+            if (colliderCache.TryGetValue(myCollider, out bool result))
```
minor nit: make else if (colliderCache.Try) 
If we clear the cache, then we know there is no point in trying to test access

julenka replied:
Done

Troy-Ferrell commented on Assets/MixedRealityToolkit/Extensions/CameraExtensions.cs:
```diff
@@ -36,10 +37,73 @@ public static bool IsInFOV(this Camera camera, Vector3 position)
+            }
+
+            // edge case: check if camera is inside the entire bounds of the collider;
+            // Consider simplifying to myCollider.bounds.Contains(CameraCache.main.transform.position)
```
Seems like left over comment?

julenka replied:
Done

julenka commented:
> Lastly, this change somewhat begs the question if there should a IsInFOVCached version? The cone is better performance

I'm not sure I understand -- are you suggesting creating a cached version of IsInFOV? I could make it, though not sure it is needed since nobody is using it right now.

julenka commented:
/azp run

julenka commented:
/azp run

julenka commented:
Troy found a bug -- If you have two cameras camera1 and camera2, if collider A is in camera1 FOV but not in camera2 FOV and we do the following:

```
camera1.IsInFovConeCached() // returns true
camera2.IsInFovConeCached() // will also return true 
```
`camera2.IsInFovConeCached()` should not return true, but it will because the cache is shared.

julenka commented:
Hi @Troy-Ferrell I fixed the edge case you though of (multiple cameras causing incorrect collider cache hits) and added a test to verify my fix. Could you please take one more look?

julenka commented:
/azp run

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Materials/HolographicButtonContentCage.mat:
```diff
@@ -175,7 +176,7 @@ Material:
     - _ClippingBorderColor: {r: 1, g: 0.2, b: 0, a: 1}
     - _ClippingPlaneBorderColor: {r: 1, g: 0.2, b: 0, a: 1}
-    - _Color: {r: 0.1254902, g: 0.1254902, b: 0.1254902, a: 1}
+    - _Color: {r: 0.3254902, g: 0.3254902, b: 0.3254902, a: 1}
```
Just double check this was intentional

julenka replied:
Good catch! Will revert.

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs:
```diff
@@ -183,11 +203,17 @@ private bool FindClosestTouchableForLayerMask(LayerMask layerMask, out BaseNearI
+                var touchable = collider.GetComponent<BaseNearInteractionTouchable>();
                 if (touchable)
                 {
+                    if (ignoreCollidersNotInFOV && !mainCam.IsInFOVConeCached(collider))
```
nit: Feel free to ignore this for now. But should use IgnoreCollidersNotInFOV public property instead of private, to be consistent with access to private variables in case the getter/setter does particular code check etcs

julenka replied:
Fixed.

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs:
```diff
@@ -62,14 +111,38 @@ public void TearDown()
+            for (int i = 0; i < TestColliders.Count; i++)
+            {
+                var test = TestColliders[i];
+                Assert.AreEqual(test.ShouldBeInFOVCamera1, testCamera.IsInFOVConeCached(test.Collider), $"TestCollider[{i}] did not match");
```
Why is the testCamera done twice but not the second camera?

julenka replied:
Good catch, I was testing that the cache was getting hit. There should be only one call to testCamera.IsInFOVConeCached

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/EditModeTests/Core/Extensions/CameraExtensionTests.cs:
```diff
@@ -62,14 +111,38 @@ public void TearDown()
+
+
+        [Test]
+        public void TestIsInFOVConeCached()
```
nit: Test comments explaining what tests do

julenka replied:
Done

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs:
```diff
@@ -100,52 +100,76 @@ public IEnumerator TestLinePointers()
-        [UnityTest]
-        public IEnumerator TestSpherePointerInsideGrabbable()
+
+        private IEnumerator TestPointerViewOfViewHelper(IMixedRealityPointer myPointer, GameObject cube, TestHand testHand)
```
Can you rename this? Not understanding viewofview

Troy-Ferrell replied:
maybe meant FieldOfView?

julenka replied:
Was supposed to be TestPointerFieldOfViewHelper

julenka replied:
Was supposed to be TestPointerFieldOfViewHelper

Troy-Ferrell commented on Assets/MixedRealityToolkit/Utilities/CameraFOVChecker.cs:
```diff
@@ -0,0 +1,85 @@
+            Collider myCollider)
+        {
+            // if the collider's size is zero, it is not visible. Return false.
+            if(myCollider.bounds.size == Vector3.zero || myCollider.transform.localScale == Vector3.zero)
```
```suggestion
            if (myCollider.bounds.size == Vector3.zero || myCollider.transform.localScale == Vector3.zero)
```

Troy-Ferrell replied:
btw, should these be floating point margin of error comparison instead of exact comparison to zero?

julenka replied:
I assumed no, since people might want to set something invisible by setting it exactly to Vector3.zero. Somebody might have a collider whose scale is very small but not zero, to represent say a point. Didn't want to ignore that.

julenka commented:
/azp run

julenka commented:
Do not merge until 2.3 stabilization branch is unlocked

julenka commented:
/azp run

davidkline-ms commented:
/azp run

davidkline-ms commented:
@julenka,

This PR is failing TestPoitnerFieldOfView and TestToggleEvents. Can you investigate?

julenka commented:
> @julenka,
> 
> This PR is failing TestPoitnerFieldOfView and TestToggleEvents. Can you investigate?

Yes, I'm aware. I am able to reproduce locally and was investigating already. I haven't understood what's wrong yet, and haven't had time to look at this since last Wednedsay. This should not block release. I will have time to look at it today, but no guarantees I will fix it. If we are close to releasing please let me know and I'll rebase off of mrtk_development.

julenka commented:
/azp run

julenka commented:
Completing, will keep an eye on MRTK_CI in case this change causes intermittent test failures.
